### ISSUE TITLE
Add an option to override the logger

### ISSUE THREAD

**Issue by domoritz** [vega/vega#2660]
I wish there was an easy way to override the default logger. Specifically, I want to redirect the logs and show them to the user in the Vega editor. 

In Vega-Lite, one can pass a logger to the compile method (https://vega.github.io/vega-lite/usage/compile.html#logging).

I am proposing a similar API, where we can pass a logger to the parser and view constructor.

**Comment by jheer** [comment]
You can already call view.logger to set the logger instance. It’s part of the Dataflow subclass. Looks like we need to add it to the View API docs. Did you also want a View constructor option? (I shied away from this in the past to avoid possible confusion with the logLevel option, but I’m open to adding it.)

**Comment by domoritz** [comment]
Oh, I missed that (and so did @marcprux).

If the logger is part of the view, then can there ever be warnings logged by the parser?

**Comment by domoritz** [comment]
Hmm, what's the API the logger needs to implement?

I tried setting the logger in the editor in https://github.com/vega/editor/pull/709 but I'm getting a bunch of errors. 

<img width="607" alt="Screen Shot 2020-05-27 at 23 07 31" src="https://user-images.githubusercontent.com/589034/83105104-dd6d5c80-a06e-11ea-8dc4-76f69598ad55.png">

I can try again once with some docs. I don't think we need a view constructor option, although I like to specify everything upfront.

**Comment by jheer** [comment]
Hmm... should follow the vega-util logger API. You can look at Dataflow.js (vega-dataflow) to see how it is invoked.

And the parser does not use logging. It just raises errors when necessary.

**Comment by jheer** [comment]
Here are the logger API docs: https://vega.github.io/vega/docs/api/util/#logger

Looks like your implementation might be missing the level method?

**Comment by domoritz** [comment]
> Looks like your implementation might be missing the level method?

It's here: https://github.com/vega/editor/pull/709/files#diff-b3176b170df0f273d8dd6ffa7768da00

**Comment by jheer** [comment]
Hmm... will investigate.

**Comment by jheer** [comment]
@domoritz I just added the following test case, which passes without issue. Also double-checked the View class, and it does not make any modifications to what the Dataflow subclass does.

```js
tape('Dataflow accepts new logger', function(t) {
  const store = [];
  const logger = {
    level: lvl => store.push('level:'+lvl),
    error: msg => store.push(msg),
    warn: msg => store.push(msg),
    info: msg => store.push(msg),
    debug: msg => store.push(msg)
  };
  const df = new vega.Dataflow().logger(logger);

  df.logLevel(99);
  df.error('error');
  df.warn('warn');
  df.info('info');
  df.debug('debug');

  t.deepEqual(store, ['level:99', 'error', 'warn', 'info', 'debug']);
  t.end();
});
```

Instead I think the error arises because [Dataflow invokes the log methods using itself as the `this` context](https://github.com/vega/vega/blob/master/packages/vega-dataflow/src/dataflow/Dataflow.js#L134) (`this._log[method].apply(this, arguments)`). The logger methods are interpreted as methods of the Dataflow class. Looks like that is why you are getting errors when your custom logger attempts to push onto an array. If you use arrow function members in your logger (or closure scoping, or `bind`, etc) you can get around this.

Also, turns out I forgot that the [View constructor already accepts a logger option](https://github.com/vega/vega/blob/master/packages/vega-view/src/View.js#L47). I will include that with the documentation update.

**Comment by domoritz** [comment]
I'm about to push a documentation pull request with documentation for the logger as well. I just need to resolve some typing issues.

**Comment by domoritz** [comment]
Pull request is at https://github.com/vega/vega/pull/2664. 

> The logger methods are interpreted as methods of the Dataflow class.

And here I was hoping that using classes would avoid `this` issues, haha. 

Just wondering, why are logging methods on the dataflow?

Also, I see that the logger has a log level but the view also takes log level as an initialization parameter. How are the two values resolved?

**Comment by domoritz** [comment]
Ahh, found it in the code "Note that by default the log level of the new logger will be used; use the logLevel method to adjust the log level as needed.".

**Comment by domoritz** [comment]
The editor now comes with Vega logging.

**Comment by jheer** [comment]
Fixed in v5.13.0 branch.
